pyxem/signals/segments.py

```diff
@@ -0,0 +1,443 @@
+# This file is part of pyXem.
+#
+# pyXem is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# pyXem is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with pyXem.  If not, see <http://www.gnu.org/licenses/>.
+"""Signal classes for segmentation results obtained from
+i) machine learning (LearningSegment) and ii) VDF images (VDFSegment).
+
+"""
+
+import numpy as np
+from tqdm import tqdm
+
+from hyperspy.signals import Signal2D
+
+from pyxem.utils.segment_utils import (norm_cross_corr, separate_watershed,
+                                       get_gaussian2d)
+from pyxem.signals.diffraction_vectors import DiffractionVectors
+from pyxem.signals.electron_diffraction2d import ElectronDiffraction2D
+from pyxem.signals import transfer_signal_axes
+
+from traits.trait_errors import TraitError
+
+
+class LearningSegment:
+    _signal_type = "learning_segment"
+
+    def __init__(self, factors, loadings):
+        # Factors as Signal2D
+        self.factors = factors
+        # Corresponding loadings as Signal2D
+        self.loadings = loadings
+
+    def correlate_learning_segments(self, corr_th_factors=0.4,
+                                    corr_th_loadings=0.4):
+        """Iterates through the factors and loadings and calculates the
+        normalized cross-correlation between all factors and all
+        loadings. Factors and loadings are summed if the correlations of
+        both factors and loadings exceed the given thresholds.
+
+        Parameters
+        ----------
+        corr_th_factors : float
+            Correlation threshold for factors. Must be between -1 and 1.
+            Factors and loadings are summed if both factors and loadings
+            have normalized cross-correlations above corr_th_factors and
+            corr_th_loadings, respectively.
+        corr_th_loadings : int, optional
+            Correlation threshold for loadings. Must be between -1 and 1.
+
+        Returns
+        -------
+        learning_segment : LearningSegment
+            LearningSegment where possibly some factors and loadings
+            have been summed.
+        """
+        # If a mask was used during the decomposition, the factors and/or
+        # loadings will contain nan, which must be converted to numbers prior
+        # to the correlations calculations.
+        factors = self.factors.map(lambda x: np.nan_to_num(x), inplace=False)
+        loadings = self.loadings.map(lambda x: np.nan_to_num(x), inplace=False)
+        factors = factors.copy().data
+        loadings = loadings.copy().data
```

Comment:
This should be doable in a neater way.

Reply:
At a glance: `self.factors.map(np.nan_to_num, inplace=False)` looks like it would work.

Reply:
Will implement.